### DISCUSSION TITLE
Return Prawn PDF block to controller

### DISCUSSION THREAD

**Discussion by aximuseng** [prawnpdf/prawn#1184] [Q&A]
In my model I have a simple method:

```
      def pdf
        pdf = Prawn::Document.new
        pdf.text 'Hello World'
        pdf
      end

```
And in my controller:

```
      format.pdf do
        send_data @model.pdf.render, filename: "model_#{@model.name}.pdf", type: "application/pdf", disposition: "inline"
      end

```
This works fine.  However I can see myself tiring of constantly having to type:

```
pdf.text 'etc..'
pdf.text 'more text'

```
vs the block method:

```
    Prawn::Document.generate("hello.pdf") do
      text "Hello World!"
    end

```
From what I see then the local scope is lost and it seem that that block returns the rendered PDF as I just get errors in my controller if I try this method.  Already getting tired of constantly typing `pdf.` before everything.

Can I modify my existing code to accomplish this?

**Comment by pointlessone** [comment]
You might want to take a look at the [View](https://prawnpdf.org/api-docs/2.3.0/Prawn/View.html) extension.

You'd also have a bit more separation of concerns with it.

```ruby
class FancyModel
  # ...
end

class FancyModelPDF
  include Prawn::View

  def initialize(model)
    @model = model
  end

  def render
    text 'etc..'
    text 'more text'

    super()
  end
end
```

And in the controller:
```ruby
format.pdf do
  send_data FancyModelPDF.new(@model).render, filename: "model_#{@model.name}.pdf", type: "application/pdf", disposition: "inline"
end
```

**Reply by halo** [reply]
I would like to emphasize what @pointlessone recommended. I've been using the implicit mode in the first years, because it was easy to type and all the documentation pointed into that direction. But for the past 5 years or so I've been using the explicit mode where you have to type `pdf.` all the time. And I would not go back to the implicit mode.

The explicit mode makes it easier for other people to understand your code if they are not very familiar with Prawn. Colleagues, beginners and consultants, for example, will be much faster in reading your code and helping out with it. Not to mention your future-you :)

The main reason for that is that it is impossible to know whether you're calling an internal Prawn method, or some other method in your code. From your example, the method `text` could be defined by you somewhere in the same file. Even if somebody knows Prawn well, you would never be sure which method was called. Not to mention that you could override existing internal methods.

Lastly, every Prawn document warrants its own ruby class. @aximuseng could you expand a little bit on the real-life use case?

* Either you generate a simple PDF with a few words, in that case it would not be a big hassle to write `pdf.` a few times, I imagine.
* Or you have a complex PDF, in which case you should have a dedicated ruby file for that, and then you can use the view extension if you really wanted to.

Just to mention it, if you have really complex PDFs, you will need multiple ruby files (say, one for the introduction, one for a table, one for the footer). In that case it would be unbearable to use the implicit mode. Normally you would just pass on the `pdf` object to those classes and each class amends the PDF with its content (at least I know of no other way).
